compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -473,10 +473,12 @@ fn encode_ty<'tcx>(
         //
         // (See https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-floating-point-types.)
         ty::Float(float_ty) => {
-            typeid.push(match float_ty {
-                FloatTy::F32 => 'f',
-                FloatTy::F64 => 'd',
-            });
+            match float_ty {
```

Comment:
These come from https://github.com/llvm/llvm-project/blob/fdef7952cbc26fd31d0d92a4f14dde58bc461fe9/clang/lib/AST/ItaniumMangle.cpp#L3124 and https://github.com/llvm/llvm-project/blob/fdef7952cbc26fd31d0d92a4f14dde58bc461fe9/clang/include/clang/Basic/TargetInfo.h#L767. We might need to do the PPC variant https://github.com/llvm/llvm-project/blob/fdef7952cbc26fd31d0d92a4f14dde58bc461fe9/clang/lib/Basic/Targets/PPC.h#L351

Not sure what the unsupported note at https://github.com/llvm/llvm-project/blob/fdef7952cbc26fd31d0d92a4f14dde58bc461fe9/clang/lib/AST/ItaniumMangle.cpp#L2947 means

Reply:
cc @rcvalle

Reply:
FFI safety: yes, we want them to be with `__float128`, `_Half`, and similar. The underlying representation is always the same as an `i16` or a *correct* `i128` on all targets I know of, but I do not know about PowerPC here.

That does of course mean that the layout of `f128` will differ across targets, e.g. SysZ ABI specifies 8-byte alignment (consistent with its i128) while most targets are 16.

That actually brings a good point, our `i128` is not C compatible so I should probably add `f128` to the `improper_ctypes` lint just to avoid any footguns. Our `i128` will have corrected alignment with the next version of LLVM, so the fix is only for a couple of months.

Reply:
> That actually brings a good point, our i128 is not C compatible

Also the tracking issue should list this as a blocking concern to be resolved before stabilizing `f128`.

Reply:
@rcvalle mind taking another look at this?

Reply:
My recommendation is:

* For `f128`, if it's compatible with `__float128` regardless of the underlying representation (i.e., changes between IEEE 754R 128-bit floating point type or IBM extended double depending on the target), we should always encode it as 'g' as it will be how it will also be encoded by Clang. (FWIW, it seems Clang uses IEEE 754R 128-bit floating point type instead of IBM extended double for `__float128` by default for ppc64le since May 2023 and `__ibm128` also doesn't seem to have a distinct encoding specified in the Itanium C++ ABI mangling.)
* For `f16`, if it's compatible with IEEE 754R 128-bit floating point type (`__fp16` and `_Float16` in Clang), we should always encode it as 'Dh'; if it's compatible with bfloat16 (which is unlikely?), we should always encode it as 'DF16b'. 
* The 'Dd,  'De', and 'Df' encodings are for when the Rust compiler and Clang support decimal floating point types.

tl;dr.: We should always encode `f128` as 'g', and `f16` as 'Dh'.

Reply:
(Note that this independent/dissociated to Rust mangling/demangling and a different encoding can be used for it.)